sdk/webpubsub/azure-messaging-webpubsubclient/azure/messaging/webpubsubclient/_client.py

```diff
@@ -772,63 +911,96 @@ def on_close(_: Any, close_status_code: int, close_msg: str):
+                first_time = True
                 while (time.time() - recovery_start < _RECOVERY_TIMEOUT) and not self._is_stopping:
                     try:
-                        self._connect(recovery_url)
+                        if ws_instance.recover_start_time is not None or not first_time:
+                            time.sleep(_RECOVERY_RETRY_INTERVAL)
+                        self._connect(recovery_url, None, recovery_start)
+                        if self._ws:
+                            self._ws.clear_reconnect_recover()
                         _LOGGER.debug("Recovery succeeded")
                         return
-                    except:  # pylint: disable=bare-except
+                    except RecoverError:
+                        _LOGGER.debug("Fail to recover, and will try in another thread")
+                        return
+                    except Exception as e:  # pylint: disable=broad-except
+                        first_time = False
+                        _LOGGER.debug("Fail to recover: %s", e)
                         _LOGGER.debug("Try to recover after %d seconds", _RECOVERY_RETRY_INTERVAL)
-                        time.sleep(_RECOVERY_RETRY_INTERVAL)
 
                 _LOGGER.warning("Recovery attempts failed after 30 seconds or the client is stopping")
-                self._handle_connection_close_and_no_recovery()
+                self._handle_connection_close_and_no_recovery(ws_instance.reconnect_tried_times)
+            else:
+                with ws_instance.cv:
+                    if close_status_code or close_msg:
+                        ws_instance.close_event = CloseEvent(
+                            close_status_code=close_status_code, close_reason=close_msg
+                        )
+                    ws_instance.cv.notify()
+
+        def on_error(ws_instance: WebSocketAppSync, error: Exception):
+            _LOGGER.warning("error happened when trying to connect: %s", error)
+            ws_instance.error_happened = error
 
         if self._is_stopping:
             raise OpenClientError("Can't open a client during closing")
 
-        self._ws = websocket.WebSocketApp(
+        # when reconnect or recovery, _thread will come here then stop by itself so don't join it
+        self._threads_join([self._thread_seq_ack])
+
+        self._ws = WebSocketAppSync(
             url=url,
             on_open=on_open,
             on_message=on_message,
             on_close=on_close,
+            on_error=on_error,
             subprotocols=[self._protocol.name] if self._protocol else [],
             header={_USER_AGENT: format_user_agent(self._user_agent)},
+            reconnect_tried_times=reconnect_tried_times,
+            recover_start_time=recover_start_time,
         )
 
         # set thread to start listen to server
-        self._thread = threading.Thread(target=self._ws.run_forever, daemon=True)
+        self._thread = threading.Thread(target=self._ws.run_forever)
```

Comment:
Why was `daemon=True` removed?

Reply:
context: https://github.com/Azure/azure-sdk-for-python/pull/34764#discussion_r1568265617
In recover/reconnect logic, old background thread `A` will stop after creating the new background thread/task `AA`. So if set `daemon`, `AA` will also stop after `A`.

Reply:
is there a way to move the creation of a new thread on recovery/reconnect back in the main thread. This way you can run them as daemons and avoid the cascading problem you described above ?

Things will close when the main thread exits

Reply:
After looking up [official doc](https://docs.python.org/3/library/threading.html#threading.Thread.daemon) about daemon thread, I realize my understanding that "thread child will be terminated if set daemon=True after thread parent done" is wrong. It is `The entire Python program exits when no alive non-daemon threads are left.` So we shall mark all the thread created by client as daemon thread.

Reply:
that is correct @msyyc , but any new threads or tasks should be created from the main thread. This will make things far easier when closing or needing clean up.

Reply:
main thread may be blocked to wait for ack message after sending and only background thread can receive ack message  from server then "wake up" main thread to complete sending. Also, only background thread can receive close message from server then trigger logic to recover/reconnect. 
First beta has adopted current design since I can't think of another solution. I will feel happy if there is other feasible solution. But if it means huge change for current design, I advise we keep current design to catch up release date for OpenAI then optimize the design in the future. (@zackliu has conveyed their urge for several times)